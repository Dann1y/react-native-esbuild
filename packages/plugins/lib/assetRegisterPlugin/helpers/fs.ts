import fs from 'node:fs/promises';
import path from 'node:path';
import type { PluginContext } from '@react-native-esbuild/core';
import type { Asset } from '@react-native-esbuild/internal';
import { logger } from '../../shared';
import {
  getAndroidAssetDestinationPath,
  getAssetDestinationPath,
  resolveAssetPath,
} from './path';

/**
 * Copy assets to platform specified destination.
 *
 * @see {@link https://github.com/react-native-community/cli/blob/v11.3.6/packages/cli-plugin-metro/src/commands/bundle/assetPathUtils.ts}
 */
export const copyAssetsToDestination = async (
  context: PluginContext,
  assets: Asset[],
): Promise<void> => {
  const { assetsDir, mode } = context;
  if (mode === 'watch') return;

  if (!assetsDir) {
    logger.warn('asset destination is not set');
    return;
  }

  logger.debug('copy assets to target destination');

  const mkdirWithAssertPath = (targetPath: string): Promise<void> => {
    const dirname = path.dirname(targetPath);
    return fs
      .access(dirname)
      .catch(() => fs.mkdir(dirname, { recursive: true }))
      .then(() => void 0);
  };

  return Promise.all(
    assets.map((asset): Promise<void> => {
      return Promise.all(
        asset.scales.map(async (scale): Promise<void> => {
          if (context.platform !== 'android') {
            const from = await resolveAssetPath(asset, scale);
            const to = path.join(
              assetsDir,
              getAssetDestinationPath(asset, scale),
            );
            await mkdirWithAssertPath(to);
            return fs.copyFile(from, to);
          }

          const from = await resolveAssetPath(asset, scale);
          const to = path.join(
            assetsDir,
            getAndroidAssetDestinationPath(asset, scale),
          );
          await mkdirWithAssertPath(to);
          await fs.copyFile(from, to);
        }),
      ).then(() => void 0);
    }),
  ).then(() => void 0);
};
