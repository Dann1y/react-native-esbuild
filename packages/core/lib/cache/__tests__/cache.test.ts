import fs from 'node:fs';
import { faker } from '@faker-js/faker';
import { CacheManager } from '..';
import type { Cache } from '../../types';

describe('CacheManager', () => {
  let manager: CacheManager;
  let mockedFs: Record<string, string>;

  beforeAll(() => {
    jest.spyOn(fs, 'mkdirSync').mockImplementation(() => void 0);
    jest
      .spyOn(fs.promises, 'readFile')
      .mockImplementation((path: string): Promise<string> => {
        const file = mockedFs[path];
        return file ? Promise.resolve(file) : Promise.reject();
      });
    jest
      .spyOn(fs.promises, 'writeFile')
      .mockImplementation((path: string, data: string): Promise<void> => {
        mockedFs[path] = data;
        return Promise.resolve();
      });
    jest.spyOn(fs.promises, 'rm').mockReturnValue(Promise.resolve());

    manager = new CacheManager();
    mockedFs = {};
  });

  describe('getCacheDirectory', () => {
    it('should match snapshot', () => {
      expect(CacheManager.getCacheDirectory()).toMatchSnapshot();
    });

    describe('when filename is present', () => {
      let filename: string;

      beforeEach(() => {
        filename = faker.system.fileName();
      });

      it('should join path with cache directory', () => {
        expect(CacheManager.getCacheDirectory(filename)).toMatch(
          new RegExp(`/${filename}$`),
        );
      });
    });
  });

  describe('memory caching', () => {
    describe('when write cache to memory', () => {
      let cacheKey: string;
      let cache: Cache;

      beforeEach(() => {
        cacheKey = faker.string.uuid();
        cache = {
          data: faker.string.alphanumeric(10),
          modifiedAt: faker.date.past().getTime(),
        };
        manager.writeToMemory(cacheKey, cache);
      });

      it('should cache data store to memory', () => {
        expect(manager.readFromMemory(cacheKey)).toMatchObject(cache);
      });

      describe('when write cache with exist key to memory', () => {
        let otherData: string;

        beforeEach(() => {
          otherData = faker.string.alphanumeric(10);
          manager.writeToMemory(cacheKey, { ...cache, data: otherData });
        });

        it('should overwrite exist cache data', () => {
          expect(manager.readFromMemory(cacheKey)).not.toMatchObject(cache);
          expect(manager.readFromMemory(cacheKey)).toMatchObject({
            ...cache,
            data: otherData,
          });
        });
      });
    });
  });

  describe('file system caching', () => {
    let hash: string;
    let data: string;

    beforeEach(async () => {
      const hashParam = { _: faker.string.alphanumeric(10) };
      hash = manager.getCacheHash(hashParam);
      data = faker.string.alphanumeric(10);
      await manager.writeToFileSystem(hash, data);
    });

    describe('when write data to file system', () => {
      it('should store data to file system', async () => {
        expect(await manager.readFromFileSystem(hash)).toEqual(data);
      });
    });
  });
});
