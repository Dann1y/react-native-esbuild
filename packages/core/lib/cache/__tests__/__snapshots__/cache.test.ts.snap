// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CacheManager getCacheDirectory should match snapshot 1`] = `"/var/folders/fp/5dmhdqcd0f178fv2jkypxx0r0000gn/T/react-native-esbuild"`;
